### PR TITLE
refactor(rustc_middle): Substs -> GenericArg

### LINKED ISSUE #110793: update internal terminology: `Substs` -> `GenericArgs`
change substitutions to the more familiar name of generic arguments in the compiler.
This includes but is not limited to:

```[tasklist]
- [x] rename `SubstsRef` and `InternalSubsts` to `GenericArgsRef<'tcx>` and `GenericArgs<'tcx>`. It is an open question and may need another MCP, but should we remove the `GenericArgsRef` alias?
- [x] rename variables and fields currently using `substs` to `args`.
- [x] update the module name of `ty::subst` to `ty::generic_args` or sth. While you're at it, probably make that module private and publicly reexport its content in the `ty` module.
- [x] rename `EarlyBinder::subst(_identity)` to `EarlyBinder::instantiate(_identity)`.
- [x] types called `[a-zA-Z]+Substs` should be renamed to `XArgs` or `XGenericArgs`.
- [x] functions containing `substs` should also use `generic_args` or `args`.
```

cc #110254 @oli-obk

### PR DESCRIPTION
**Pull request by mdibaiee**
resolves #110793

- [x] rename `SubstsRef` and `InternalSubsts` to `GenericArgsRef<'tcx>` and `GenericArgs<'tcx>`.
- [x] rename variables and fields currently using `substs` to `args`.
- [x] update the module name of `ty::subst` to `ty::generic_args` or sth. Make that module private and publicly reexport its content in the ty module.
- [x] rename `EarlyBinder::subst(_identity)` to `EarlyBinder::instantiate(_identity)`.
- [x] types called `[a-zA-Z]+Substs` renamed to `XArgs`.
- [x] functions containing `substs` now use `args` or `generic_args` (mostly the former).

However, the verb of "substituting" is still being used here and there, mostly in comments. I think that can be a separate PR as part of https://github.com/rust-lang/rust/issues/110254 to change the verb to `replace_generics` or something similar.

### CODE REVIEW

**Comment by mdibaiee** [comment]
r? ghost

**Review by oli-obk — APPROVED** [review]

**Comment by BoxyUwU** [comment]
I hate to show up and bikeshed an mcp accepted name on a PR thats _already_ been approved (although the mcp was literaly 2 years ago so I don't think its unreasonable to act like maybe it would have been good to revisit this a bit). I really do not like the name `GenericArgsRef`. It's longer than `SubstsRef`, the `Ref` doesnt add any value (when is anyone using the non-ref version anyway). `GenericArgs` is atleast Roughly the same size but still larger, I'd suggest `ty::Args` but that its about generic args and I'd also worry that `ty::GenArgs` miht be read as "generator args".

**Comment by BoxyUwU** [comment]
maybe after this lands we can rename `GenericArgsRef` and `GenericArgs` again? :grimacing:

**Comment by mdibaiee** [comment]
@BoxyUwU removing `GenericArgsRef` is actually noted in the original issue here https://github.com/rust-lang/rust/issues/110793 as a "maybe". 

I don't have anything against the proposal to replace `GenericArgsRef` with actual references to `GenericArgs`, but note that this won't be _shorter_. The actual reference is 3 characters longer than the alias 😬
```
GenericArgsRef<'tcx>
&'tcx GenericArgs<'tcx>
```

If there is consensus that it's a better idea to get rid of `GenericArgsRef`, I'm happy to make a PR for that too

**Comment by BoxyUwU** [comment]
I am not suggesting to replace `GenericsArgsRef<'tcx>` with the actual type. I'm suggesting to rename `GenericArgs` to `GenericArgsInner<'tcx>` and ` rename `GenericArgsRef<'tcx>` to `GenericArgs<'tcx>` since its the most commonly used of the two.

**Comment by mdibaiee** [comment]
@BoxyUwU ah okay, I mixed up your suggestion with the one in the original issue, so this is a separate suggestion. I think this could actually work well.

I still would like to get this PR merged first since it can cause a lot of conflicts, so better get it out of the way, then replacing the type name is easier (this PR does a bunch of other stuff).

**Comment by spastorino** [comment]
If we are going to rename as @BoxyUwU suggested, I'd personally think we should do it in this PR to avoid an interim of a name we know is gonna be replaced. Also, please @mdibaiee add the corresponding change to the Rustc Dev Guide.

**Comment by nnethercote** [comment]
> I'm suggesting to rename `GenericArgs` to `GenericArgsInner<'tcx>` and `rename`GenericArgsRef<'tcx>`to`GenericArgs<'tcx>` since its the most commonly used of the two.

Yes, Huffman encoding is a good rule of thumb: more common names should be shorter.

**Comment by lcnr** [comment]
> If we are going to rename as @BoxyUwU suggested, I'd personally think we should do it in this PR to avoid an interim of a name we know is gonna be replaced. Also, please @mdibaiee add the corresponding change to the Rustc Dev Guide.

I disagree here and strongly believe we should merge this PR as is. And will approve this PR once it is rebased.

I don't know if and and how I want us to change replace `GenericsArgsRef` later on. I am unhappy with simply renaming `InternalSubsts` to `GenericArgsInner` and `SubstsRef` to `GenericArgs`. Doing this change by itself is also far easier and won't be as merge conflict heavy. If we change something "more fundamentally" about this, I would like a more invasive change here which should not happen in a general renaming PR.

The general move from `Subst` to `GenericArg` is a major improvement and I think `GenericArgsRef` is still better than `SubstsRef`. We have already had multiple rounds of bikeshedding here and this has been discussed since pretty much forever: the original compiler MCP was in June of 2021. I do not want us to block this PR for a minor improvement while taking the risk for nothing to happen again for a while.

File: src/tools/cargo
Comment (oli-obk) [inline src/tools/cargo:1]: you accidentally got a submodule bump in here.

File: src/tools/tidy/src/deps.rs
Comment (oli-obk) [inline src/tools/tidy/src/deps.rs:1]: this change seems accidental, too

**Pull request merged**
